[jayhee3] iP

jytheawesome opened the pull request:
## Welcome to the brand new release of Duke! 
> One should not fear failing. Instead, one should fear not being able to stand up and try again after failing. " - A wise man

## What is Duke? 

Duke is a friendly chat-bot that can help you schedule your tasks. 

Ever wanted to plan tasks as fast as your brain can think of them?
Are you tired of dealing with pesky calendar apps that are frustrating and slow to use with unintuitive user interfaces? 

Fret not! Duke is here to save the day!

- Purely text-based interface so you can plan and schedule tasks **_at light speed_** :zap:
- Simple to use user commands with **_no need for headache and frustration_** :relaxed: 
- **_Easily_** keep track of all your _**tasks, events, deadlines**_! :alarm_clock:
- Last of all, it is **_free_**! 

## Trying it out 
Here are the steps: 

1. Download the file over [here](https://github.com/jayhee3/ip/releases)
2.  Click on `Assets`
3. Download the `duke.jar` file to your Desktop
4.  Run the `duke.jar` file and observe how Duke can help make your life easier and more productive! 

## Coming soon: 

- [ ] Reminders for upcoming tasks and deadlines
- [x] Ability to add recurring tasks (that repeat daily, weekly, or monthly)   


## A note for advanced users: 

If you wish to run the `duke.jar` file from the terminal, you can do so by running the following command in the directory where the `duke.jar` file is downloaded within the terminal: 

``` java
java -jar duke.jar
```
#
![DALL·E 2023-02-04 16 17 49 - A cute robot that everyone will love that helps you schedule tasks](https://user-images.githubusercontent.com/95480800/216757061-58adf39c-9118-4d09-bbaa-1a08574dd867.png)

Commit:
Add Gradle support

jytheawesome committed:
Increment 1 functionalities: Greets, echos, and says goodbye to user.

jytheawesome committed:
Increment 2; Added list feature, add to list feature

jytheawesome committed:
Added features for Increment 3, ability to mark and unmark Tasks, Tasks are now encapsulated as a class. TaskStorage is now also a class.

jytheawesome committed:
Added features for Increment 4;
Added 3 new classes, Deadline, ToDo, Event which all inherit Task.

jytheawesome committed:
Initiated automated text testing, added DukeException class to prepare for next increment.

jytheawesome committed:
increment 5 features. Handled a large variety of possible errors.

jytheawesome committed:
deleted DukeException as it is not needed for now.

jytheawesome committed:
Completed changes for increment 6. Added the ability to delete tasks. Also updated the status message for "mark" and "unmark" commands so users know which task the action was performed on.

yyj-02 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,134 @@
     }
+
+
+    public static void oneWordCommand(String userInput, TaskStorage taskStorage) throws Exception {
```
I think the method name `oneWordCommand` is a bit ambiguous, can we choose a method name that starts with an action verb? In my opinion, this can make it more intuitive for other developers. Suggestions: `handleUserInput()`

yyj-02 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,134 @@
+        if (userInput.equals("list")) {
+            taskStorage.listTasks();
+            return;
+        }
```
Can we use a `switch ... case ...` using `firstWord` as the switch?

yyj-02 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,134 @@
+                    + "\nYou now currently have "
+                    + taskStorage.getStorageCount() + " tasks.");
+
+        } if (firstWord.equals("delete")) {
```
I think we should not stack `if` statements like this. Can we use `else if` instead?

yyj-02 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,134 @@
+        } if (firstWord.equals("delete")) {
+            String bodyMessage = userInput.split(" ", 2)[1];
+            int taskNumber = Integer.parseInt(bodyMessage);
+            Task toDelete = taskStorage.getTask(taskNumber);
```
Can we make the variable name more intuitive, i.e. `taskToDelete`

yyj-02 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,134 @@
+                    + taskStorage.getStorageCount() + " tasks.");
+            return;
+
+
```
I think there shouldn't be extra spaces after a return statement. Can we remove these line breaks throughout the files?

Other examples in other parts of the repository:

* do not need line break after {
* do not need line break before }
* avoid double line break (only one is enough)

jytheawesome replied:
Thanks for the comment regarding line breaks. According to the java coding standard, the documentation regarding the use of blank lines is not the most detailed or specific, so I often am unsure if a blank line is needed or not.

Have taken note of your suggestions of instances where blank lines are not needed. In your opinion and experience coding, when/where do you usually make use of blank lines to improve readability? @yyj-02

yyj-02 replied:
Oops I just saw this, usually, I put blanks to separate and categorise the logics. So if this part of the code deals with validating inputs, I will not put space between them. But after this section, if for example the next section is about interpreting this inputs, I will put a space between these 2 sections

yyj-02 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,134 @@
+
+
+            respond("We have removed this task: " + toDelete.provideDetails() + "\nYou now have "
+                                + taskStorage.getStorageCount() + " tasks remaining");
```
Can we correct the indentation level here?

yyj-02 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    @Override
+    public String provideDetails() {
```
I feel that `provideDetails()` is not that intuitive to users, why did you choose not to `@Override` the default `toString()` method in Object class? I think that will be a better design, feel free to correct me.

jytheawesome replied:
That is a very good suggestion. Thanks for suggesting, I will implement `toString()` instead. Thanks @yyj-02.

yyj-02 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Status of the Task.
+     */
+    boolean completed = false;
```
Can we change this to `isCompleted`? This makes it follows the convention more

yyj-02 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,36 @@
+     * Provides Details of the Task.
+     * @return String detail message of Task.
+     */
+    public String provideDetails() {
+        return completed ? "[x] " + task
+                : "[ ] " + task;
+    }
```
Can we override the default `toString` method instead?

yyj-02 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Flips the status of the Task.
+     */
+    public void updateTask() {
```
If the method is to flip the status, can we change the method name to `toggleStatus` to make it more self explanatory?

yyj-02 commented on src/main/java/TaskStorage.java:
```diff
@@ -0,0 +1,59 @@
+     * Updates the status of a Task.
+     * @param number The number representing the task to be updated.
+     */
+    public void updateTask(int number) {
```
Can we change the parameter name? The parameter name can be `index` to make it more self explanatory, similar for the rest below.

jytheawesome replied:
> Good effort 😊 Here is a some general overview:
> 
> * Good effort to separate the different tasks into separate components
> * Can try to choose method/variable names that are self-explanatory
> * Avoid unnecessary line breaks/indentations
> 
> Moving forward, you can also try to extract the different components into their own class, i.e. `Parser`, `Command`, `Ui`
> 
> All the best! 🔥

Thank you for your suggestions on using better names for variables according to Java conventions, the appropriate use of blank lines, and also suggesting the use of switch-case instead of if-else statements. My code at this instance was not well-refined yet. Thanks again for the comments!

yyj-02 reviewed (commented):
Good effort 😊 Here is a some general overview:

* Good effort to separate the different tasks into separate components
* Can try to choose method/variable names that are self-explanatory
* Avoid unnecessary line breaks/indentations

Moving forward, you can also try to extract the different components into their own class, i.e. `Parser`, `Command`, `Ui`

All the best! 🔥

jytheawesome committed:
Add ability to store tasks in hard drive so they do not disappear when Duke session is closed.

jytheawesome committed:
Make DeadLine class store and print deadlines as a LocalDate

jytheawesome committed:
Merge branch 'branch-Level-7'

jytheawesome committed:
Merge branch-8

jytheawesome committed:
Add more OOP by separating out the Ui, Storage, and TaskList classes.

jytheawesome committed:
Add all existing classes into a package called duke.

thennant commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,144 @@
+                    ui.respond("I have added this new task:\n" + task.provideDetails()
+                            + "\nYou now currently have "
+                            + tasks.getTaskCount() + " tasks.");
+
+                } else if (firstWord.equals("deadline")) {
+                    //Rest of message describes the Task.
+                    String body = userInput.split(" ", 2)[1];
+                    DeadLine newTask = new DeadLine("deadline", body, false);
+                    tasks.addTask(newTask);
+                    storage.addTask(newTask.getStorageLine());
+                    ui.respond("I have added this new task:\n" + newTask.provideDetails()
+                            + "\nYou now currently have "
+                            + tasks.getTaskCount() + " tasks.");
+
+                } else if (firstWord.equals("event")) {
+                    //Rest of message describes the Task.
+                    String body = userInput.split(" ", 2)[1];
+                    Event newTask = new Event("event", body, false);
+                    tasks.addTask(newTask);
+                    storage.addTask(newTask.getStorageLine());
+
+                    ui.respond("I have added this new task:\n" + newTask.provideDetails()
+                            + "\nYou now currently have "
+                            + tasks.getTaskCount() + " tasks.");
+
+
+                } else if (firstWord.equals("delete")) {
+                    //second word should be an integer
+                    String secondWord = userInput.split(" ", 2)[1];
+                    int taskNumber = Integer.parseInt(secondWord);
+                    Task task = tasks.getTask(taskNumber);
+                    tasks.deleteTask(taskNumber);
+                    storage.deleteTask(task.getStorageLine());
+
+                    ui.respond("We have removed this task: " + task.provideDetails() + "\nYou now have "
+                            + tasks.getTaskCount() + " tasks remaining");
+
+                } else {
+                    ui.showCommandError();
+                }
```
Maybe consider refactoring the code here to be stored into the parser class?

jytheawesome replied:
Thanks! I missed out the Parser class while completing this increment. @thennant

thennant commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Lists all the tasks stored.
+     */
+    public void listTasks() {
+        String topDivider = "~~~~~~~~~~~~~~~~o~~~~~~~~~~~~~~~~\n" + "Duke's Response: \n"
+                    + "Here are the tasks in your list: \n";
+        String botDivider = "~~~~~~~~~~~~~~~~o~~~~~~~~~~~~~~~~";
+        System.out.println(topDivider);
```
Should this be stored under the Ui class?

jytheawesome replied:
Yes, that is a good suggestion. Have done that already. @thennant

thennant commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,15 @@
+package duke;
+class ToDo extends Task {
+
+    public ToDo(String keyword, String message, Boolean completed) {
+        super(keyword, message, completed);
+    }
```
good naming conventions!

thennant commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,193 @@
+package duke;
+import java.io.File;
+import java.io.FileWriter;
+import java.util.Scanner;
+import java.util.ArrayList;
```
Might want to consider breaking up the imports into groups (ie. leave a space after line 3) to follow the coding standards?

thennant commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,193 @@
+     */
+    public void addTask(String input) throws DukeException {
+        try {
+            //Instantiating the File class
+            File myObj = new File("duke.txt");
+            //Instantiating the Scanner class to read the file
+            Scanner sc = new Scanner(myObj);
+
+            StringBuffer buffer = new StringBuffer();
+
+            //Reading lines of the file and appending them to StringBuffer
+            while (sc.hasNextLine()) {
+                buffer.append(sc.nextLine() + "\n");
+            }
+            String fileContents = buffer.toString();
```
Seems like theres multiple repeition with calling the files here, you might want to consider storing it under a single method and check for the task type instead?

thennant reviewed (commented):
Hi there!
Overall there was good use of OOP principles and naming conventions in your code. You might want to consider refactoring the code in your run() method to be stored in a parser class instead. Additionally, you might also want to consider packaging your classes more such as moving your event, task, deadline and todo under a single 'Task' package.

jytheawesome committed:
Add Parser Class.

jytheawesome committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

jytheawesome committed:
Add JUnit tests for ToDo String Description and Parse behaviour for "bye" and invalid command message.

jytheawesome committed:
Remove "updated hard drive" message.

jytheawesome committed:
Add javadoc comments

jytheawesome committed:
Add style changes so it follows coding standards

jytheawesome committed:
add find feature

jytheawesome committed:
Merge branch 'branch-A-JavaDoc'

jytheawesome committed:
Merge branch 'branch-A-CodingStandard'

jytheawesome committed:
Merge branch 'branch-Level-9'

jytheawesome commented:
> Hi there! Overall there was good use of OOP principles and naming conventions in your code. You might want to consider refactoring the code in your run() method to be stored in a parser class instead. Additionally, you might also want to consider packaging your classes more such as moving your event, task, deadline and todo under a single 'Task' package.

Thanks for the suggestions! The recommendation to package classes more finely is a good one.

jytheawesome committed:
Add graphical user interface to Duke

jytheawesome committed:
Change the version of javafx api used

jytheawesome committed:
Add assert functionality

jytheawesome committed:
Add the use of java assertations

There are key junctures in our code that are likely to throw a
critical error.

Using java assertations helps us to catch these critical faults
early.

Let's add java assertations in our code.

jytheawesome committed:
Merge pull request #2 from jayhee3/branch-A-Assertions

Add the use of java assertations

jytheawesome committed:
Improve the quality of the code through refactoring

The current code has much room for improvement in terms of clarity
and length.

Let's improve the code quality.

jytheawesome committed:
Merge pull request #3 from jayhee3/branch-A-CodeQuality

Improve the quality of the code through refactoring

jytheawesome committed:
Clean up code

jytheawesome committed:
Modify text-ui-test files

jytheawesome committed:
Add DoAfter task functionality

Currently, no tasks support the capability for the user to complete the task after a certain date.

This would be useful for the user to keep track of tasks that are to be done only after a certain date.

Let's add this task functionality to our app.

jytheawesome committed:
Merge branch 'branch-B-DoAfterTasks'

jytheawesome committed:
Add more Exception handling

jytheawesome committed:
Increase GUI minimum height

jytheawesome committed:
Revert "Increase GUI minimum height"

This reverts commit 10ce9b57e338d38a0bcecf7a90bbb04baf7a5591.

jytheawesome committed:
Reduce minimum GUI height

jytheawesome committed:
Add more javadoc comments

jytheawesome committed:
Modify user and bot image

jytheawesome committed:
Update README.md

jytheawesome committed:
Fix bug for storing do-after tasks in hard drive

jytheawesome committed:
Update README.md

jytheawesome committed:
Update README.md

jytheawesome committed:
Bug fix DoAfter

jytheawesome committed:
Add greeting message for GUI

jytheawesome committed:
Add background color

jytheawesome committed:
Recover test files

jytheawesome committed:
Remove Duke String

jytheawesome committed:
Rectify bye command

jytheawesome committed:
Add test code
Improve GUI

jytheawesome committed:
Delete Ui.png

jytheawesome committed:
Add Ui.png

jytheawesome committed:
Update README.md

jytheawesome committed:
Fix GUI bugs

jytheawesome committed:
Merge branch 'master' of https://github.com/jayhee3/ip

jytheawesome committed:
Update README.md
